src/lib.rs

```diff
@@ -67,46 +67,134 @@ pub use attr::{AttributeValue, AttributeValueParts, Attributes};
 
 type CowStr<'s> = std::borrow::Cow<'s, str>;
 
+/// A trait for rendering [`Event`]s to an output format.
+///
+/// The output can be written to either a [`std::fmt::Write`] or a [`std::io::Write`] object.
+///
+/// If ownership of the [`Event`]s cannot be given to the renderer, use [`Render::push_ref`] or
+/// [`Render::write_ref`].
+///
+/// An implementor needs to at least implement the [`Render::render_event`] function that renders a
+/// single event to the output. If anything needs to be rendered at the beginning or end of the
+/// output, the [`Render::render_prologue`] and [`Render::render_epilogue`] can be implemented as
+/// well.
+///
+/// # Examples
+///
+/// Push to a [`String`] (implements [`std::fmt::Write`]):
+///
+/// ```
+/// # use jotdown::Render;
+/// # let events = std::iter::empty();
+/// let mut output = String::new();
+/// let mut renderer = jotdown::html::Renderer::default();
+/// renderer.push(events, &mut output);
+/// ```
+///
+/// Write to standard output with buffering ([`std::io::Stdout`] implements [`std::io::Write`]):
+///
+/// ```
+/// # use jotdown::Render;
+/// # let events = std::iter::empty();
+/// let mut out = std::io::BufWriter::new(std::io::stdout());
+/// let mut renderer = jotdown::html::Renderer::default();
+/// renderer.write(events, &mut out).unwrap();
+/// ```
 pub trait Render {
-    /// Push [`Event`]s to a unicode-accepting buffer or stream.
-    fn push<'s, I: Iterator<Item = Event<'s>>, W: fmt::Write>(
-        &self,
-        events: I,
-        out: W,
-    ) -> fmt::Result;
-
-    /// Write [`Event`]s to a byte sink, encoded as UTF-8.
+    /// Render a single event.
+    fn render_event<'s, W>(&mut self, e: &Event<'s>, out: W) -> std::fmt::Result
+    where
+        W: std::fmt::Write;
+
+    /// Render something before any events have been provided.
```

Comment:
Should we add a note in the docs about the default behavior (does nothing)? I know it is somehow mentioned in the top-level docs for this trait, but people browsing the documentation might see "provided methods" and think that this does something by default.

Reply:
Maybe better to not give it a default implementation at all, and let implementor conciously set it to do nothing if nothing is needed? Might make it clearer that it is a implementation-oriented function rather than for the user?

Reply:
That would require more boilerplate for the implementor though. Probably not a huge deal either way, but I say we just leave a note in the documentation.

Reply:
added note